File: lib/face_detector.dart
import 'dart:io';

import 'package:flutter_vision/constants.dart';
import 'package:flutter_vision/flutter_vision.dart';

class FaceDetector {
  bool _hasBeenOpened = false;
  bool _isClosed = false;

  Future<bool> startDetection() async {
    assert(!_isClosed);

    _hasBeenOpened = true;
    return await FlutterVision.cameraChannel.invokeMethod<bool>(MethodNames.addFaceDetector);
  }

  Future<String> startDetectionIOS(double width, double height) async {
    assert(!_isClosed);

    _hasBeenOpened = true;
    return await FlutterVision.channel.invokeMethod<String>(
      MethodNames.addFaceDetector,
      <String, double>{
        'width': width,
        'height': height,
      },
    );
  }

  Future<bool> close() async {
    if (!_hasBeenOpened) _isClosed = true;
    if (_isClosed) return Future<bool>.value(true);

    _isClosed = true;

    if (Platform.isIOS) {
      return await FlutterVision.channel.invokeMethod<bool>(MethodNames.closeFaceDetector);
    } else {
      return await FlutterVision.cameraChannel.invokeMethod<bool>(MethodNames.closeFaceDetector);
    }
  }
}

class Face {
  final double rotY;
  final double rotZ;
  final double smile;
  final double rightEyeOpen;
  final double leftEyeOpen;
  final int trackingId;
  final Position leftEye;
  final Position rightEye;
  final Position leftCheek;
  final Position rightCheek;
  final Position leftEar;
  final Position rightEar;
  final Position mouthLeft;
  final Position mouthBottom;
  final Position mouthRight;
  final Position noseBase;
  final BoundingBox boundingBox;
  final double faceAngle;

  Face(dynamic _data)
      : rotY = _data['rotY'],
        rotZ = _data['rotZ'],
        smile = _data['smile'],
        rightEyeOpen = _data['rightEyeOpen'],
        leftEyeOpen = _data['leftEyeOpen'],
        trackingId = _data['trackingId'],
        leftEye = _data['leftEye'] != null ? Position(_data['leftEye']) : null,
        rightEye = _data['rightEye'] != null ? Position(_data['rightEye']) : null,
        leftCheek = _data['leftCheek'] != null ? Position(_data['leftCheek']) : null,
        rightCheek = _data['rightCheek'] != null ? Position(_data['rightCheek']) : null,
        leftEar = _data['leftEar'] != null ? Position(_data['leftEar']) : null,
        rightEar = _data['rightEar'] != null ? Position(_data['rightEar']) : null,
        mouthLeft = _data['mouthLeft'] != null ? Position(_data['mouthLeft']) : null,
        mouthBottom = _data['mouthBottom'] != null ? Position(_data['mouthBottom']) : null,
        mouthRight = _data['mouthRight'] != null ? Position(_data['mouthRight']) : null,
        noseBase = _data['noseBase'] != null ? Position(_data['noseBase']) : null,
        faceAngle = _data['faceAngle'],
        boundingBox = _data['boundingBox'] != null && _data['boundingBox']['top'] != null
            ? BoundingBox.fromMap(_data['boundingBox'])
            : null;

  static List<Face> fromList(List<dynamic> data) {
    return data.map((m) => Face(m)).toList();
  }
}

class Position {
  final double x;
  final double y;
  final double z;

  Position(dynamic data)
      : x = data["x"],
        y = data["y"],
        z = data["z"];
}

class BoundingBox {
  final num left;
  final num top;
  final num width;
  final num height;

  BoundingBox.fromMap(dynamic data)
      : left = data['left'],
        top = data['top'],
        width = data['width'],
        height = data['height'];

  BoundingBox.fromLTWH(this.left, this.top, this.width, this.height);
}

File: android/src/main/java/com/travelunion/flutter_vision/detectors/FaceDetector.java
package com.travelunion.flutter_vision.detectors;

import android.media.Image;

import androidx.annotation.NonNull;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.ml.vision.FirebaseVision;
import com.google.firebase.ml.vision.common.FirebaseVisionImage;
import com.google.firebase.ml.vision.common.FirebaseVisionPoint;
import com.google.firebase.ml.vision.face.FirebaseVisionFace;
import com.google.firebase.ml.vision.face.FirebaseVisionFaceDetector;
import com.google.firebase.ml.vision.face.FirebaseVisionFaceDetectorOptions;
import com.google.firebase.ml.vision.face.FirebaseVisionFaceLandmark;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicBoolean;

import io.flutter.plugin.common.EventChannel;

public class FaceDetector implements Detector {
    private final FirebaseVisionFaceDetector detector;

    public FaceDetector(FirebaseVision vision) {
        FirebaseVisionFaceDetectorOptions options =
                new FirebaseVisionFaceDetectorOptions.Builder()
                        .setLandmarkMode(FirebaseVisionFaceDetectorOptions.ALL_LANDMARKS)
                        .setClassificationMode(FirebaseVisionFaceDetectorOptions.ALL_CLASSIFICATIONS)
                        .build();

        detector = vision.getVisionFaceDetector(options);
    }

    @Override
    public void handleDetection(final FirebaseVisionImage image, final EventChannel.EventSink result, final AtomicBoolean throttle) {
        detector
                .detectInImage(image)
                .addOnSuccessListener(
                        new OnSuccessListener<List<FirebaseVisionFace>>() {
                            @Override
                            public void onSuccess(List<FirebaseVisionFace> faces) {
                                List<Map<String, Object>> data = new ArrayList<>();

                                for (FirebaseVisionFace face : faces) {
                                    Map<String, Object> faceMap = new HashMap<>();

                                    faceMap.put("rotY", face.getHeadEulerAngleY());
                                    faceMap.put("rotZ", face.getHeadEulerAngleZ());

                                    FirebaseVisionFaceLandmark leftEye = face.getLandmark(FirebaseVisionFaceLandmark.LEFT_EYE);
                                    if (leftEye != null) {
                                        faceMap.put("leftEye", getPosition(leftEye.getPosition()));
                                    }

                                    FirebaseVisionFaceLandmark rightEye = face.getLandmark(FirebaseVisionFaceLandmark.RIGHT_EYE);
                                    if (rightEye != null) {
                                        faceMap.put("rightEye", getPosition(rightEye.getPosition()));
                                    }

                                    FirebaseVisionFaceLandmark leftEar = face.getLandmark(FirebaseVisionFaceLandmark.LEFT_EAR);
                                    if (leftEar != null) {
                                        faceMap.put("leftEar", getPosition(leftEar.getPosition()));
                                    }

                                    FirebaseVisionFaceLandmark rightEar = face.getLandmark(FirebaseVisionFaceLandmark.RIGHT_EAR);
                                    if (rightEar != null) {
                                        faceMap.put("rightEar", getPosition(rightEar.getPosition()));
                                    }

                                    FirebaseVisionFaceLandmark rightCheek = face.getLandmark(FirebaseVisionFaceLandmark.RIGHT_CHEEK);
                                    if (rightCheek != null) {
                                        faceMap.put("rightCheek", getPosition(rightCheek.getPosition()));
                                    }

                                    FirebaseVisionFaceLandmark leftCheek = face.getLandmark(FirebaseVisionFaceLandmark.LEFT_CHEEK);
                                    if (leftCheek != null) {
                                        faceMap.put("leftCheek", getPosition(leftCheek.getPosition()));
                                    }

                                    FirebaseVisionFaceLandmark mouthLeft = face.getLandmark(FirebaseVisionFaceLandmark.MOUTH_LEFT);
                                    if (mouthLeft != null) {
                                        faceMap.put("mouthLeft", getPosition(mouthLeft.getPosition()));
                                    }

                                    FirebaseVisionFaceLandmark mouthBottom = face.getLandmark(FirebaseVisionFaceLandmark.MOUTH_BOTTOM);
                                    if (mouthBottom != null) {
                                        faceMap.put("mouthBottom", getPosition(mouthBottom.getPosition()));
                                    }

                                    FirebaseVisionFaceLandmark mouthRight = face.getLandmark(FirebaseVisionFaceLandmark.MOUTH_RIGHT);
                                    if (mouthRight != null) {
                                        faceMap.put("mouthRight", getPosition(mouthRight.getPosition()));
                                    }

                                    FirebaseVisionFaceLandmark noseBase = face.getLandmark(FirebaseVisionFaceLandmark.NOSE_BASE);
                                    if (noseBase != null) {
                                        faceMap.put("noseBase", getPosition(noseBase.getPosition()));
                                    }

                                    faceMap.put("smile", face.getSmilingProbability());
                                    faceMap.put("rightEyeOpen", face.getRightEyeOpenProbability());
                                    faceMap.put("leftEyeOpen", face.getLeftEyeOpenProbability());
                                    faceMap.put("trackingId", face.getTrackingId());

                                    data.add(faceMap);
                                }

                                Map<String, Object> res = new HashMap<>();
                                res.put("eventType", "faceDetection");
                                res.put("data", data);

                                throttle.set(false);

                                if(!data.isEmpty()) {
                                    result.success(res);
                                }
                            }
                        })
                .addOnFailureListener(
                        new OnFailureListener() {
                            @Override
                            public void onFailure(@NonNull Exception exception) {
                                throttle.set(false);
                                result.error("faceDetectionError", exception.getLocalizedMessage(), null);
                            }
                        });
    }

    @Override
    public void handleDetection(Image originalImage, FirebaseVisionImage image, EventChannel.EventSink eventSink, AtomicBoolean throttle) {
        this.handleDetection(image, eventSink, throttle);
    }

    @Override
    public void close() throws IOException {
        detector.close();
    }

    private Map<String, Object> getPosition(FirebaseVisionPoint position) {
        Map<String, Object> result = new HashMap<>();

        result.put("x", position.getX());
        result.put("y", position.getY());
        result.put("z", position.getZ());

        return result;
    }
}

